odoo/addons/base/models/ir_ui_view.py

```diff
@@ -747,29 +768,73 @@ def apply_inheritance_specs(self, source, specs_tree, pre_locate=lambda s: True)
         """
-        inherit_tree = collections.defaultdict(list)
-        for view in self.get_inheriting_views_arch(model):
-            inherit_tree[view.inherit_id].append(view)
-        return self._apply_view_inheritance(source, inherit_tree)
-
-    def _apply_view_inheritance(self, source, inherit_tree):
-        # recursively apply inheritance following the given inheritance tree
-        for view in inherit_tree[self]:
-            arch_tree = etree.fromstring(view.arch.encode('utf-8'))
-            if self._context.get('inherit_branding'):
-                view.inherit_branding(arch_tree)
-            source = view.apply_inheritance_specs(source, arch_tree)
-            source = view._apply_view_inheritance(source, inherit_tree)
-        return source
+        self.ensure_one()
+        assert self.mode == 'primary'
+
+        # We achieve a pre-order depth-first hierarchy traversal where
+        # primary views (and their children) are traversed after all the
+        # extensions for the current primary view have been visited.
+        #
+        # https://en.wikipedia.org/wiki/Tree_traversal#Depth-first_search_of_binary_tree
+        #
+        # Example:                  hierarchy = {
+        #                               1: [2, 3],  # primary view
+        #             1*                2: [4, 5],
+        #            / \                3: [],
+        #           2   3               4: [6],     # primary view
+        #          / \                  5: [7, 8],
+        #         4*  5                 6: [],
+        #        /   / \                7: [],
+        #       6   7   8               8: [],
+        #                           }
+        #
+        # Tree traversal order (`view` and `queue` at the `while` stmt):
+        #   1 [2, 3]
+        #   2 [5, 3, 4]
+        #   5 [7, 8, 3, 4]
+        #   7 [8, 3, 4]
+        #   8 [3, 4]
+        #   3 [4]
+        #   4 [6]
+        #   6 []
+        combined_arch = etree.fromstring(self.arch.encode('utf-8'))
+        if self.env.context.get('inherit_branding'):
+            combined_arch.attrib.update({
+                'data-oe-model': 'ir.ui.view',
+                'data-oe-id': str(self.id),
+                'data-oe-field': 'arch',
+            })
+
+        # The depth-first traversal is implemented with a double-ended queue.
+        # The queue is traversed from left to right, and after each view in the
+        # queue is processed, its children are pushed at the left of the queue,
+        # so that they are traversed in order.  The queue is therefore mostly
+        # used as a stack.  An exception is made for primary views, which are
+        # pushed at the other end of the queue, so that they are applied after
+        # all extensions have been applied.
+        queue = collections.deque(sorted(hierarchy[self], key=lambda v: v.mode))
+        while queue:
+            view = queue.popleft()
+            arch = etree.fromstring(view.arch.encode('utf-8'))
```

Comment:
Is `encode()` actually necessary?

Reply:
If it is, `'utf-8'` isn't, as it is the default value for `encode()`.

Reply:
IIRC the word "string" in "fromstring" refers to py2's string aka bytes in py3.

Reply:
> IIRC the word "string" in "fromstring" refers to py2's string aka bytes in py3.

The [lxml](https://lxml.de/api/lxml.etree-module.html#fromstring) doc is not fancy, but you could just try it, and you'll see it accepts both text and bytes, so no need for extra encoding.

```py
Python 3.8.5 (default, May 27 2021, 13:30:53) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from lxml import etree
>>> etree.fromstring("<a></a>")
<Element a at 0x7f1c894b3ec0>
```